File: frontend/src/types/loan.ts
export enum LoanStatus {
  NEW = "New",
  IN_PROCESS = "InProcess",
  CONDITIONAL_APPROVAL = "ConditionalApproval",
  APPROVED = "Approved",
  SUSPENDED = "Suspended",
  DECLINED = "Declined",
  WITHDRAWN = "Withdrawn",
  FUNDED = "Funded",
}

export enum LoanPurposeType {
  PURCHASE = "Purchase",
  REFINANCE = "Refinance",
  CASH_OUT_REFINANCE = "CashOutRefinance",
  CONSTRUCTION_TO_PERMANENT = "ConstructionToPermanent",
}

export enum LoanType {
  CONVENTIONAL = "Conventional",
  FHA = "FHA",
  VA = "VA",
  USDA = "USDA",
}

export enum UserRole {
  LOAN_OFFICER = "LoanOfficer",
  PROCESSOR = "Processor",
  UNDERWRITER = "Underwriter",
  CLOSER = "Closer",
  SECONDARY_MARKETING = "SecondaryMarketing",
  BRANCH_MANAGER = "BranchManager",
  COMPLIANCE_OFFICER = "ComplianceOfficer",
  ADMIN = "Admin",
}

export enum DocumentStatus {
  REQUESTED = "Requested",
  RECEIVED = "Received",
  REVIEWED = "Reviewed",
  ACCEPTED = "Accepted",
  REJECTED = "Rejected",
}

export interface Borrower {
  id: string;
  loan_id: string;
  borrower_classification: "Primary" | "CoBorrower";
  first_name: string;
  last_name: string;
  middle_name?: string;
  email: string;
  phone: string;
  credit_score?: number;
  created_at: string;
  updated_at: string;
}

export interface Document {
  id: string;
  loan_id: string;
  document_type: string;
  document_status: DocumentStatus;
  s3_key?: string;
  original_filename: string;
  uploaded_by_id?: string;
  reviewed_by_id?: string;
  uploaded_at?: string;
  reviewed_at?: string;
  created_at: string;
}

export interface AUSResult {
  id: string;
  loan_id: string;
  system: "DU" | "LPA";
  recommendation: string;
  risk_class: string;
  findings_summary: string;
  run_at: string;
  raw_response?: Record<string, unknown>;
}

export interface LoanListItem {
  id: string;
  loan_number: string;
  status: LoanStatus;
  loan_amount: number;
  loan_purpose_type: LoanPurposeType;
  loan_type: LoanType;
  property_city: string;
  property_state: string;
  created_at: string;
  primary_borrower_name?: string;
  assigned_lo_name?: string;
  days_in_status?: number;
}

export interface Loan {
  id: string;
  loan_number: string;
  status: LoanStatus;
  loan_purpose_type: LoanPurposeType;
  loan_type: LoanType;
  loan_amount: number;
  note_rate_percent?: number;
  property_address_line: string;
  property_city: string;
  property_state: string;
  property_zip: string;
  property_county?: string;
  created_by_id: string;
  assigned_lo_id?: string;
  assigned_processor_id?: string;
  assigned_underwriter_id?: string;
  mismo_data?: Record<string, unknown>;
  application_received_date?: string;
  estimated_close_date?: string;
  created_at: string;
  updated_at: string;
  borrowers: Borrower[];
  documents?: Document[];
}

export interface LoanCreatePayload {
  loan_purpose_type: LoanPurposeType;
  loan_type: LoanType;
  loan_amount: number;
  property_address_line: string;
  property_city: string;
  property_state: string;
  property_zip: string;
}

export interface LoanUpdatePayload {
  status?: LoanStatus;
  loan_amount?: number;
  note_rate_percent?: number;
  application_received_date?: string; // YYYY-MM-DD
  estimated_close_date?: string;
  assigned_lo_id?: string;
  assigned_processor_id?: string;
  assigned_underwriter_id?: string;
}

File: backend/app/schemas/loan.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional, List
from uuid import UUID

from pydantic import BaseModel, Field, ConfigDict, field_validator

from app.models.loan import LoanStatus, LoanPurposeType, LoanType


# ---------------------------------------------------------------------------
# Borrower sub-schema (used inside LoanResponse)
# ---------------------------------------------------------------------------

class BorrowerResponse(BaseModel):
    id: UUID
    loan_id: UUID
    borrower_classification: str
    first_name: str
    last_name: str
    middle_name: Optional[str] = None
    email: str
    phone: str
    credit_score: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Document sub-schema (used inside LoanResponse)
# ---------------------------------------------------------------------------

class DocumentResponse(BaseModel):
    id: UUID
    loan_id: UUID
    document_type: str
    document_status: str
    original_filename: str
    uploaded_at: Optional[datetime] = None
    reviewed_at: Optional[datetime] = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ---------------------------------------------------------------------------
# Loan schemas
# ---------------------------------------------------------------------------

class LoanCreate(BaseModel):
    loan_purpose_type: LoanPurposeType
    loan_type: LoanType
    loan_amount: Decimal = Field(gt=0, le=50_000_000, decimal_places=2)
    property_address_line: str = Field(min_length=5, max_length=200)
    property_city: str = Field(min_length=2, max_length=100)
    property_state: str = Field(min_length=2, max_length=2)
    property_zip: str = Field(min_length=5, max_length=10)

    @field_validator("property_state")
    @classmethod
    def uppercase_state(cls, v: str) -> str:
        return v.upper()


class LoanUpdate(BaseModel):
    """Partial update schema — all fields optional."""
    status: Optional[LoanStatus] = None
    loan_amount: Optional[Decimal] = Field(default=None, gt=0, le=50_000_000)
    note_rate_percent: Optional[Decimal] = Field(default=None, ge=0, le=30)
    property_address_line: Optional[str] = Field(default=None, min_length=5, max_length=200)
    property_city: Optional[str] = Field(default=None, min_length=2, max_length=100)
    property_state: Optional[str] = Field(default=None, min_length=2, max_length=2)
    property_zip: Optional[str] = Field(default=None, min_length=5, max_length=10)
    property_county: Optional[str] = Field(default=None, max_length=100)
    assigned_lo_id: Optional[UUID] = None
    assigned_processor_id: Optional[UUID] = None
    assigned_underwriter_id: Optional[UUID] = None
    application_received_date: Optional[date] = None
    estimated_close_date: Optional[date] = None


class LoanListItem(BaseModel):
    id: UUID
    loan_number: str
    status: LoanStatus
    loan_amount: Decimal
    loan_purpose_type: LoanPurposeType
    loan_type: LoanType
    property_city: str
    property_state: str
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class LoanResponse(BaseModel):
    id: UUID
    loan_number: str
    status: LoanStatus
    loan_purpose_type: LoanPurposeType
    loan_type: LoanType
    loan_amount: Decimal
    note_rate_percent: Optional[Decimal] = None
    property_address_line: str
    property_city: str
    property_state: str
    property_zip: str
    property_county: Optional[str] = None
    created_by_id: UUID
    assigned_lo_id: Optional[UUID] = None
    assigned_processor_id: Optional[UUID] = None
    assigned_underwriter_id: Optional[UUID] = None
    mismo_data: Optional[dict] = None
    application_received_date: Optional[date] = None
    estimated_close_date: Optional[date] = None
    created_at: datetime
    updated_at: datetime
    borrowers: List[BorrowerResponse] = []
    documents: List[DocumentResponse] = []

    model_config = ConfigDict(from_attributes=True)


class PaginatedLoanList(BaseModel):
    items: List[LoanListItem]
    total: int
    skip: int
    limit: int
